src/wallet/wallet.cpp

```diff
@@ -82,15 +82,63 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
-// Custom deleter for shared_ptr<CWallet>.
-static void ReleaseWallet(CWallet* wallet)
+static Mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cv;
+static std::set<CWallet*> g_unloading_wallet_set;
+
+// Unregister and delete the wallet right after BlockUntilSyncedToCurrentChain
+// so that it's in sync with the current chainstate.
+static void DeleteWallet(CWallet* wallet)
 {
     wallet->WalletLogPrintf("Releasing wallet\n");
     wallet->BlockUntilSyncedToCurrentChain();
     wallet->Flush();
+    UnregisterValidationInterface(wallet);
     delete wallet;
 }
 
+// Custom deleter for shared_ptr<CWallet>.
+static void ReleaseWallet(CWallet* wallet)
+{
+    bool unloading{false};
+    {
+        LOCK(g_wallet_release_mutex);
+        unloading = g_unloading_wallet_set.erase(wallet) == 1;
+    }
+    if (unloading) {
+        // Delegate wallet deletion to UnloadWallet caller, just wake it up.
+        g_wallet_release_cv.notify_all();
+    } else {
+        // UnloadWallet was not called, just delete the wallet since there's no
+        // more shared pointers to the wallet.
+        DeleteWallet(wallet);
+    }
+}
+
+void UnloadWallet(std::shared_ptr<CWallet>&& wallet)
+{
+    // Mark wallet for unloading.
+    CWallet* pwallet = wallet.get();
+    {
+        LOCK(g_wallet_release_mutex);
+        assert(g_unloading_wallet_set.insert(pwallet).second);
+    }
+    // The wallet can be in use so it's not possible to explicitly unload here.
+    // Notify the unload intent so that all remaining shared pointers are
+    // released.
+    pwallet->NotifyUnload();
```

Comment:
Could we `assert(NotifyUnload.empty())` in the wallet destructor to verify delete isn't happening too early and clients are fully detached?

Reply:
Will try.